### PR TITLE
[RFC] screen: avoid artifacts

### LINKED ISSUE #8357: after `inoremap jk <esc>` `jk` leaves artifact on buffer
- `nvim --version`: HEAD
- Vim (version: ) behaves differently?
- Operating system/version: Ubuntu 18.04
- Terminal name/version: st
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC

```

### Actual behaviour

After adding this binding `inoremap jk <esc>`, `jk` leaves an artifact (the character `k`) on the screen.

### Expected behaviour

After `jk` I shouldn't see anything on the buffer.

### PR DESCRIPTION
**Pull request by mhinz**
Put back the condition that was accidentally removed in https://github.com/neovim/neovim/commit/d42f934bcb3e9e876e5e7ba0ab5cd824175fd10c.

```diff
-    if (enc_utf8 && ScreenLinesUC[off] != 0)
-      bytes[utfc_char2bytes(off, bytes)] = NUL;
-    else if (enc_dbcs == DBCS_JPNU && ScreenLines[off] == 0x8e) {
-      bytes[0] = ScreenLines[off];
-      bytes[1] = ScreenLines2[off];
-      bytes[2] = NUL;
-    } else if (enc_dbcs && MB_BYTE2LEN(bytes[0]) > 1) {
-      bytes[1] = ScreenLines[off + 1];
-      bytes[2] = NUL;
-    }
+    bytes[utfc_char2bytes(off, bytes)] = NUL;
```

Fixes #8357

### CODE REVIEW

**Comment by jamessan** [comment]
Is there a test we can add for this?

**Comment by bfredl** [comment]
Can also reproduce it with `<c-r><esc>` in insert mode, but there must be characters after the cursor, otherwise the tempoary `"` is never put in the first place.

**Comment by mhinz** [comment]
Not sure how to test it. I tried this naive screen test:

```lua
command('inoremap jk <esc>')
feed('ifoojk')
screen:snapshot_util()
```

This shows just `foo` instead of `fooj`, so I guess the screen is redrawn at some point before `screen:snapshot_util()`.

**Comment by bfredl** [comment]
@mhinz You will need to add a `snapshot_util()`/`except()` between the j and the k, otherwise the j won't get displayed in the first case. (in insert mode nvim will try to merge multiple chars present in the input buffer)

**Comment by mhinz** [comment]
@bfredl Thanks for the explanation!

I used two commits for easier checking whether the test actually works or not. (It does! 🎉) Feel free to squash.

File: test/functional/ui/screen_basic_spec.lua [hunk 1]
```diff
@@ -6,6 +6,7 @@ local insert = helpers.insert
 local eq = helpers.eq
 local eval = helpers.eval
 local iswin = helpers.iswin
+local wait = helpers.wait
```
Comment (jamessan) [inline test/functional/ui/screen_basic_spec.lua]: Unnecessary import

Comment (mhinz) [reply test/functional/ui/screen_basic_spec.lua]: Yup, I just noted that...

File: test/functional/ui/screen_basic_spec.lua [hunk 2]
```diff
@@ -645,4 +645,43 @@ describe('Screen', function()
       ]])
     end)
   end)
+
+  -- Regression test for #8357
+  it('does not have artifacts', function()
```
[the remark this answers is not in the document]
Comment (mhinz) [reply test/functional/ui/screen_basic_spec.lua]: I don't like it myself, but I can't come up with a better one that is not long enough. I'm open for suggestions.

[the remark this answers is not in the document]
Comment (bfredl) [reply test/functional/ui/screen_basic_spec.lua]: "does not have artifacts after temporary chars in insert mode" doesn't sound to long to me.

**Pull request merged**